messagingpush/src/main/java/io/customer/messagingpush/processor/PushMessageProcessorImpl.kt

```diff
@@ -0,0 +1,88 @@
+) : PushMessageProcessor {
+
+    /**
+     * Responsible for storing and updating recent messages in queue atomically.
+     * Once this method is called, the current implementation marks the notification
+     * as processed and future calls for same [deliveryId] will receive true indicating
+     * that the notification is processed already.
+     *
+     * @param deliveryId unique message id received in push payload.
+     * @return true if the message was processed previously; false otherwise.
+     * Callers should generally process notifications only if false was returned.
+     */
+    @Synchronized
+    @VisibleForTesting
+    internal fun getOrUpdateMessageAlreadyProcessed(deliveryId: String?): Boolean {
+        when {
+            deliveryId.isNullOrBlank() -> {
+                // Ignore messages with empty/invalid deliveryId
+                logger.debug("Received message with empty deliveryId")
+                return true
+            }
+
+            PushMessageProcessor.recentMessagesQueue.contains(deliveryId) -> {
+                // Ignore messages that were processed already
+                logger.debug("Received duplicate message with deliveryId: $deliveryId")
+                return true
+            }
+
+            else -> {
+                // Else add this deliveryId to the queue and assure queue don't exhaust its capacity
+                if (PushMessageProcessor.recentMessagesQueue.size >= PushMessageProcessor.RECENT_MESSAGES_MAX_SIZE) {
+                    PushMessageProcessor.recentMessagesQueue.removeLast()
+                }
+                PushMessageProcessor.recentMessagesQueue.addFirst(deliveryId)
+                logger.debug("Received new message with deliveryId: $deliveryId")
+                return false
+            }
+        }
+    }
+
+    override fun processGCMMessageIntent(intent: Intent) {
+        val bundle = intent.extras
+        val deliveryId = bundle?.getString(PushTrackingUtil.DELIVERY_ID_KEY)
+        val deliveryToken = bundle?.getString(PushTrackingUtil.DELIVERY_TOKEN_KEY)
+        // Not a CIO push notification
+        if (deliveryId == null || deliveryToken == null) return
+
+        if (!getOrUpdateMessageAlreadyProcessed(deliveryId = deliveryId)) {
+            // We only track delivered metrics from GCM right now
+            trackDeliveredMetrics(deliveryId, deliveryToken)
+        }
+    }
+
+    override fun processRemoteMessageDeliveredMetrics(deliveryId: String, deliveryToken: String) {
+        if (!getOrUpdateMessageAlreadyProcessed(deliveryId = deliveryId)) {
+            // We only track delivered metrics here for Firebase messages as the caller
+            // processes and displays the notification already and FCM guarantees to send
+            // callbacks only once for each notification, so duplication is not possible within
+            // Firebase message receivers
+            trackDeliveredMetrics(deliveryId, deliveryToken)
+        }
+    }
```

Comment:
The implementation of these 2 functions is almost identical except `processGCMMessageIntent` parses an `Intent` while the other doesn't. 

When you look at the caller of `processRemoteMessageDeliveredMetrics` in `CustomerIOPushNotificationHandler.kt`, an `Intent` is parsed before calling `processRemoteMessageDeliveredMetrics`. 

Can we refactor the code to avoid this code duplication?

Reply:
They look slightly identical at the moment but they both have different purpose. They should remain 2 separate methods because:

`processGCMMessageIntent` is the called directly from GCM receiver and is planned to display notification as well for wrapper SDKs in future

`processRemoteMessageDeliveredMetrics` was initially designed to receive `RemoteMessage` and parse delivery parameters by itself to make it easier for the caller. But since `deliveryId` and `deliveryToken` are being used by `CustomerIOPushNotificationHandler` to determine whether it is a CIO push notification or not, and also to generate push payload, moving them out of `CustomerIOPushNotificationHandler` was resulting in even more confusions.

So, out of all the solutions I tried before, the current structure looked the best one to me. I would stick to current changes for now. We can then gradually move more changes to processor and utilize it better.